package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,333 @@
+  log_debug("client_config: sbp_sender_id=%d\n", config->sbp_sender_id);
+  log_debug("client_config: enabled=%d\n", config->enabled);
+}
+
+/**
+ *  Settings and HTTP client configuration setup.
+ */
+
+// TODO (mookerji): Implementing settings!
+RC setup_settings(skylark_client_config_t* config)
+{
+  /* TYPE_SKYLARK_MODE \ */
+  /*   = settings_type_register_enum(skylark_mode_enum, &skylark_mode_settings_type); */
+  /* READ_ONLY_PARAMETER('skylark', */
+  /*                     'enable', */
+  /*                     config->enabled, */
+  /*                     TYPE_SKYLARK_MODE); */
+  /* READ_ONLY_PARAMETER(SETTINGS_SKYLARK_GROUP, */
+  /*                     SETTINGS_SKYLARK_URL, */
+  /*                     config->endpoint_url, */
+  /*                     TYPE_STRING); */
+  return NO_ERROR;
+}
+
+RC get_uuid_channel(skylark_client_config_t* config)
+{
+  (void)config;
+  return -E_NOT_IMPLEMENTED;
+}
+
+RC get_device_uuid(skylark_client_config_t* config)
+{
+  strcpy(config->device_uuid, DEFAULT_DEVICE_UID);
+  return NO_ERROR;
+}
+
+RC get_broker_endpoint(skylark_client_config_t* config)
+{
+  strcpy(config->endpoint_url, DEFAULT_BROKER_ENDPOINT);
+  return NO_ERROR;
+}
+
+RC get_accept_header(skylark_client_config_t* config)
+{
+  strcpy(config->accept_type_header, SBP_V2_ACCEPT_TYPE);
+  return NO_ERROR;
+}
+
+RC get_content_header(skylark_client_config_t* config)
+{
+  strcpy(config->content_type_header, SBP_V2_CONTENT_TYPE);
+  return NO_ERROR;
+}
+
+RC format_device_header(skylark_client_config_t* config)
+{
+  sprintf(config->device_header, DEVICE_UID_HEADER_FMT, config->device_uuid);
+  return NO_ERROR;
+}
+
+RC get_sbp_sender_id(u16* sender_id)
+{
+  *sender_id = SBP_SENDER_ID;
+  char sbp_sender_id_string[32];
+  if (file_read_string("/cfg/sbp_sender_id", sbp_sender_id_string,
+                       sizeof(sbp_sender_id_string)) < 0) {
+    return E_GENERIC_ERROR;
+
+  }
+  *sender_id = strtoul(sbp_sender_id_string, NULL, 10);
+  return NO_ERROR;
+}
+
+RC get_config(skylark_client_config_t* config)
+{
+  RC rc = NO_ERROR;
+  if ((rc = get_broker_endpoint(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_accept_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_content_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  strcpy(config->encoding, STREAM_ENCODING);
+  if ((rc = get_device_uuid(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = format_device_header(config)) < NO_ERROR) {
+    return rc;
+  }
+  if ((rc = get_sbp_sender_id(&config->sbp_sender_id)) < NO_ERROR) {
+    return rc;
+  }
+  return rc;
+}
+
+/**
+ *  Download (i.e., rover) processes.
+ */
+
+// Callback used by libcurl to pass data read from Skylark. Writes to pipe.
+//
+// TODO (mookerji): replace
+static size_t download_callback(void *p, size_t size, size_t n, void *up)
+{
+  int *fd = (int *)up;
+  ssize_t m = write(*fd, p, size * n);
+  return m;
+}
```

Comment:
Sorry about my offline comments - I think it's fine to keep these callbacks next to the curl routines because they're super general - just write to a file descriptor.